docs/language/operators/README.md

```diff
@@ -301,6 +302,326 @@ conn  2018-03-24T17:15:20.607695Z CpjMvj2Cvj048u6bF1 10.164.94.120 39169     10.
+field from the right-hand input to appear in the joined results.
+
+The Zed script `right-join.zed`:
+
+```mdtest-input right-join.zed
+from (
+  file fruit.ndjson => sort flavor;
+  file people.ndjson => sort likes;
+) | right join on flavor=likes fruit:=name
+```
+
+Executing the Zed script:
+
+```mdtest-command
+zq -z -I right-join.zed
+```
+
+#### Output:
+```mdtest-output
+{name:"jessie",age:30,likes:"plain",fruit:"figs"}
+{name:"quinn",age:14,likes:"sweet",note:"many kids enjoy sweets",fruit:"banana"}
+{name:"quinn",age:14,likes:"sweet",note:"many kids enjoy sweets",fruit:"strawberry"}
+{name:"quinn",age:14,likes:"sweet",note:"many kids enjoy sweets",fruit:"dates"}
+{name:"morgan",age:61,likes:"tart",fruit:"apple"}
+{name:"chris",age:47,likes:"tart",fruit:"apple"}
+```
+
+#### Example #4 - Inputs from Pools
+
+As our prior examples all used `zq`, we used `file` in our `from()` block to
+pull our respective inputs from named file sources. However, if the inputs are
+stored in Pools in a Zed lake, the Pool names would instead be specified in the
+`from()` block.
+
+Here we'll load our input data to Pools in a temporary Zed Lake, then execute
+our inner join using `zed lake query`. If the Zed Lake had been fronted by a
+`zed lake serve` process, the equivalent operations would be performed over the
+network via `zed api`.
+
+Notice that because we happened to use `-orderby` to sort our Pools by the same
+keys that we reference in our `join`, we did not need to use any explicit
+upstream `sort`.
+
+The Zed script `inner-join-pools.zed`:
+
+```mdtest-input inner-join-pools.zed
+from (
+  fruit => sort flavor;
+  people => sort likes;
+) | inner join on flavor=likes eater:=name
+```
+
+Populating the Pools, then executing the Zed script:
+
+```mdtest-command
+ZED_LAKE_ROOT=`mktemp -d`
+zed lake init -q
+zed lake create -q -p fruit -orderby flavor:asc
+zed lake create -q -p people -orderby likes:asc
+zed lake load -q -p fruit fruit.ndjson
+zed lake load -q -p people people.ndjson
+zed lake query -z -I inner-join-pools.zed
+```
+
+#### Output:
+```mdtest-output
+{name:"figs",color:"brown",flavor:"plain",eater:"jessie"}
+{name:"dates",color:"brown",flavor:"sweet",note:"in season",eater:"quinn"}
+{name:"banana",color:"yellow",flavor:"sweet",eater:"quinn"}
+{name:"strawberry",color:"red",flavor:"sweet",eater:"quinn"}
+{name:"apple",color:"red",flavor:"tart",eater:"chris"}
+{name:"apple",color:"red",flavor:"tart",eater:"morgan"}
+```
+
+#### Example #5 - Streamed input
+
+In addition to the named files and Pools like we've used in the prior examples,
+Zed is also intended to work on streams of data. Here we'll combine our file
+sources into a stream that we'll pipe into `zq` via stdin. Because join requires
+two separate inputs, here we'll use the `has()` function to identify the
+records in the stream that will be treated as the left and right sides.
+
+The Zed script `inner-join-streamed.zed`:
+
+```mdtest-input inner-join-streamed.zed
+* | split (
```

Comment:
If I don't include the leading `* |`, I get an error:

```
no input specified: use a command-line file or a Zed source operator
```

Since we've got the convention elsewhere in the language that the user can lead off the leading `* |` if they want all the data, could we make a change to allow this with `split` here? If so, I'm happy to open an issue.

Reply:
This happens because `zq` implements special behavior for `split` when it's the first operation in the query. In that case, `zq` expects its command line to have exactly one file per `split` branch and feeds the first file to the first branch and so on.

We should remove that special behavior eventually, though I'm not sure we're ready to do so now. Let's discuss with @mccanne.

Until then, `switch` avoids `* |`.
```
switch (
  case has(color) => sort flavor
  case has(age) => sort likes
) | inner join on flavor=likes eater:=name
```

Reply:
@nwt: Cool, thanks for that guidance. I've pushed a commit to instead use `switch` for now since it gets us to the same place and seems less confusing. As we discussed as a group, I'm game to see the requirement for the leading `* |` go away by opting for `from file` as the way users would be expected to wire up separate file inputs to branches of the flowgraph.